pkg/master/reconcilers/endpointsadapter.go

```diff
@@ -48,8 +48,14 @@ func NewEndpointsAdapter(endpointClient corev1client.EndpointsGetter, endpointSl
 
 // Get takes the name and namespace of the Endpoints resource, and returns a
 // corresponding Endpoints object if it exists, and an error if there is any.
+// If an Endpoints object is found and an endpointSliceClient exists, a
+// matching EndpointSlice will be created or updated.
 func (adapter *EndpointsAdapter) Get(namespace, name string, getOpts metav1.GetOptions) (*corev1.Endpoints, error) {
-	return adapter.endpointClient.Endpoints(namespace).Get(name, getOpts)
+	endpoints, err := adapter.endpointClient.Endpoints(namespace).Get(name, getOpts)
+	if err == nil && adapter.endpointSliceClient != nil {
+		_, err = adapter.ensureEndpointSliceFromEndpoints(namespace, endpoints)
```

Comment:
Doing a write in a Get method doesn't seem proper. I'd expect these to be written in the same place we persist the Endpoints object. 

Also, is endpointSliceClient when the feature is disabled?

Reply:
Thanks for reviewing.

The problem is when the Endpoints object is persisted, the EndpointSlice feature might not be enabled yet, like bootstrapping a cluster with kubeadm (which doesn't have a feature gate for endpointslice) or upgrading from a version that doesn't have EndpointSlice.  
Then after enabling this feature, when EndpointReconciler reconciles the endpoint, it will skip the creation as it finds an existing one. The Create method which is supposed to sync Endpoints and EndpointSlice gets no chance to run. 

Since the EndpointsAdapter is in charge of reading and writing both Endpoints and EndpointSlices and keeping reconciler EndpointSlices agnostic, I thought it might be better to not break it so make Endpoints and EndpointSlices consistent in Get method in the same way as Create/Update methods do. Otherwise the EndpointReconciler might need to be adjusted to call epAdaptor.Create even if epAdaptor.Get finds one. https://github.com/kubernetes/kubernetes/blob/362e17ff9fdc2639de7cc465bfea2150d7d56a40/pkg/master/reconcilers/lease.go#L162-L170

endpointSliceClient is nil when the feature is disabled, the added code follows the way in which Create/Update method decides whether to sync EndpointSlice.
https://github.com/kubernetes/kubernetes/blob/362e17ff9fdc2639de7cc465bfea2150d7d56a40/pkg/master/master.go#L240-L244

Please let me know if the explanation makes sense to you. If not, could you suggest a proper way to fix it?

Reply:
I hadn't thought that through, but you're completely right that a `Get` method really shouldn't be writing anything. An approach with the least required change would be to rename the method, but it's hard to come up with a reasonable name - `GetEndpointsAndEnsureEndpointSlices` seems like it would be too long. 

What might make more sense would be to modify each of the reconcilers to call `r.epAdapter.ensureEndpointSliceFromEndpoints(namespace, endpoints)` at the end of their `ReconcileEndpoints` methods. That would require adding a check inside that method to ensure an `endpointSliceClient` was set on the adapter, but otherwise would be relatively straightforward I think.

Reply:
> What might make more sense would be to modify each of the reconcilers to call `r.epAdapter.ensureEndpointSliceFromEndpoints(namespace, endpoints)` at the end of their `ReconcileEndpoints` methods. That would require adding a check inside that method to ensure an `endpointSliceClient` was set on the adapter, but otherwise would be relatively straightforward I think.
> 
I tried not to break the current EndpointAdaptor's effort that minimizes change to reconcilers and keeps it EndpointSlice agnostic. If adding`r.epAdapter.ensureEndpointSliceFromEndpoints` to the end of `ReconcileEndpoints` methods, I'm not sure whether it still makes sense to keep the adapter because calling `ensureEndpointSliceFromEndpoints` in `r.epAdapter.Create` and `r.epAdapter.Update` would be unnecessary (because it might get called twice each time). There could just be a ensureEndpointSliceFromEndpoints call after reconciling the endpoint no matter the endpoint is created/update/not changed.

To keep the change small, I updated the method name to `GetAndSync`, does it make sense to you? Or you would prefer adding this explicit call `ensureEndpointSliceFromEndpoints` to [doReconcile](https://github.com/kubernetes/kubernetes/blob/8d8a0684866d75aa195de20871a6164794b11ebc/pkg/master/reconcilers/lease.go#L162)?

Reply:
we should not change this from a getter to a getter+writer. syncing from endpoints to endpointslices in the getter means that a single invocation of doReconcile will complete with the two resources not matching (and the endpointslice won't be corrected until the **next** invocation of doReconcile, which is not what we want).

Reply:
> If addingr.epAdapter.ensureEndpointSliceFromEndpoints to the end of ReconcileEndpoints methods, I'm not sure whether it still makes sense to keep the adapter

yeah... unfortunately, a lot of the logic in doReconcile makes assumptions about the endpoint implementation (the short-circuiting logic assumes that only a single resource is being managed, for example).

> There could just be a ensureEndpointSliceFromEndpoints call after reconciling the endpoint no matter the endpoint is created/update/not changed.

calling ensureEndpointSliceFromEndpoints() just before all non-error returns from doReconcile (even short-circuits) seems like it could work.

Reply:
I have updated as you suggested, PTAL. Thanks!